### DISCUSSION TITLE
Viewer based on Label instead of Protocol

### DISCUSSION THREAD

**Discussion by tinohager** [BinaryKits/ZPLUtility#120] [General]
The Viewer project is historically based on the Label project. We have already prepared the Protocol project, but now we need to port the Viewer to the other project. From my point of view, this step should be completed first before adding new features.

What is your opinion on this @ingfdoaguirre?

**Comment by primo-ppcg** [comment]
I've begun work on this. It seems to me that the entire `Viewer.CommandAnalyzers` namespace is redundant with the `Protocol.Commands` namespace, in particular with the methods `IsCommandParsable` and `ParseCommand`. The `ParseCommand` methods appear to be more robustly written, anyway.

They aren't, however, particularly convenient to work with. `IsCommandParsable` is an instance method, and `ParseCommand` is a mutator. To parse a list of previously separated command strings:

1. Construct a list of `CommandBase` instances for each type of command
2. For each command string:
a. Scan the list of command instances for a parser
b. Get a `ConstructorInfo` for the command type using reflection
c. Invoke the constructor
d. Call `ParseCommand` on the newly created instance

Any thoughts?

**Reply by tinohager** [reply]
I think it will go in this direction

**Reply by primo-ppcg** [reply]
More concretely, I'm asking if there is any intention/openness to change the CommandBase interface to be more convenient for parsing a pre-existing command string. For example, a static method that returns an instance of the correct type for any abitrary command. Currently, it's necessary to know the command type before it can be parsed. And then, it can only be parsed by an instance of that explicit type, which is then mutated in place.

**Reply by tinohager** [reply]
I think the `BinaryKits.Zpl.Protocol` project can still be changed in all directions it was primarily about reading, understanding and validating the Zpl language. Since this project is not yet in use it can be adapted in all directions.

**Reply by primo-ppcg** [reply]
@tinohager I've created a branch for this project: https://github.com/BinaryKits/BinaryKits.Zpl/tree/ProtocolViewer

You can see from the test cases the direction I've taken it:

```csharp
[TestMethod]
public void ParseCommand_ValidCommand1_Successful()
{
    var command = CommandBase.ParseCommand("^BY2,2.9,10");
    Assert.IsTrue(command is BarCodeFieldDefaultCommand);
    if (command is BarCodeFieldDefaultCommand barcodeCommand)
    {
        Assert.AreEqual(2, barcodeCommand.ModuleWidth);
        Assert.AreEqual(2.9, barcodeCommand.WideBarToNarrowBarWidthRatio);
        Assert.AreEqual(10, barcodeCommand.BarCodeHeight);
    }
}
```

The idea being that any command can be parsed without needing to know beforehand what it is.

**Reply by tinohager** [reply]
@primo-ppcg looks good for me
